### PR TITLE
Added two indexes on sales_flat_order and sales_flat_order_item tables

### PR DESCRIPTION
**Pull request by fballiano**
This solution was already discussed and agreed in issue https://github.com/OpenMage/magento-lts/issues/384 I just created the PR into the Mage_Sales module since in 2022 I don't see the need for a new dedicated module for this specific task (as discussed in the previously linked issue).

### CODE REVIEW

**Review by tobihille — DISMISSED** [review]

**Review by ADDISON74 — APPROVED** [review]

**Review by tobihille — DISMISSED** [review]

**Review by ADDISON74 — APPROVED** [review]

**Comment by fballiano** [comment]
tested on 1million sales_flat_order_items rows on my local machine and took only a few seconds (10ish, also considering that cache was cold) so I guess it will not create troubles even to big stores in production.

**Review by ADDISON74 — APPROVED** [review]

**Review by ADDISON74 — DISMISSED** [review]

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
+ *
+ * @category    Mage
+ * @package     Mage_Sales
+ * @copyright  Copyright (c) 2006-2020 Magento, Inc. (http://www.magento.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+/* @var Mage_Sales_Model_Entity_Setup $installer */
```
Comment (sreichel) [inline app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:21]:
```suggestion
/** @var Mage_Sales_Model_Entity_Setup $installer */
```

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
+ *
+ * @category    Mage
+ * @package     Mage_Sales
```
Comment (sreichel) [inline app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:16]:
```suggestion
 * @category   Mage
 * @package    Mage_Sales
```

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        ADD INDEX `IDX_SALES_FLAT_ORDER_CUSTOMER_EMAIL` (`customer_email` (15));
+    ");
+}
+
+// Add index to sales_flat_order_item.product_id for fast join/lookup
+$this->getConnection()->addIndex(
+    $installer->getTable('sales/order_item'),
+    'IDX_SALES_FLAT_ORDER_ITEM_PRODUCT_ID',
+    ['product_id']
+);
```
Comment (sreichel) [inline app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:40]:
```suggestion
$keyList = $installer->getConnection()->getIndexList($installer->getTable('sales/order_item'));
if (!isset($keyList['IDX_SALES_FLAT_ORDER_ITEM_PRODUCT_ID'])) {
    $this->getConnection()->addIndex(
        $installer->getTable('sales/order_item'),
        'IDX_SALES_FLAT_ORDER_ITEM_PRODUCT_ID',
        ['product_id']
    );
}
```

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+$installer->startSetup();
+
+// Add index to sales_flat_order on customer_email for fast lookup, only first 15 bytes
+$keyList = $installer->getConnection()->getIndexList($installer->getTable('sales/order'));
+if (!isset($keyList[strtoupper('customer_email')])) {
+    $installer->run("
+        ALTER TABLE {$installer->getTable('sales/order')}
+        ADD INDEX `IDX_SALES_FLAT_ORDER_CUSTOMER_EMAIL` (`customer_email` (15));
+    ");
+}
```
Comment (sreichel) [inline app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:33]:
```suggestion
if (!isset($keyList['IDX_SALES_FLAT_ORDER_CUSTOMER_EMAIL'])) {
    $installer->run("
        ALTER TABLE {$installer->getTable('sales/order')}
        ADD INDEX `IDX_SALES_FLAT_ORDER_CUSTOMER_EMAIL` (`customer_email` (15));
    ");
}
```

What is the correct syntax for using ```(`customer_email` (15))``` with `$this->getConnection()->addIndex()`)

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
+ *
+ * @category    Mage
+ * @package     Mage_Sales
+ * @copyright  Copyright (c) 2006-2020 Magento, Inc. (http://www.magento.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
```
Comment (sreichel) [inline app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:18]:
```suggestion
 * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
```

File: app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php (hunk 3 above)
Comment (sreichel) [reply app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php:40]: This seems not requierd.

**Review by sreichel — CHANGES_REQUESTED** [review]
Please check my suggestion ... when i reset version in `core_resource` table an run update again ...

```
Error in file: "/var/www/html/app/code/core/Mage/Sales/sql/sales_setup/upgrade-1.6.0.10-1.6.0.11.php" - SQLSTATE[42000]: Syntax error or access violation: 1061 Duplicate key name 'IDX_SALES_FLAT_ORDER_CUSTOMER_EMAIL'
```

**Comment by fballiano** [comment]
it should be ok now

**Review by sreichel — APPROVED** [review]

**Review by ADDISON74 — APPROVED** [review]

**Pull request merged**